internal/runners/uploadingredient/uploadingredient.go

```diff
@@ -0,0 +1,343 @@
+	"github.com/ActiveState/cli/internal/rtutils/p"
+	"github.com/ActiveState/cli/pkg/platform/api"
+	graphModel "github.com/ActiveState/cli/pkg/platform/api/graphql/model"
+	"github.com/ActiveState/cli/pkg/platform/api/graphql/request"
+	auth "github.com/ActiveState/cli/pkg/platform/authentication"
+	"github.com/ActiveState/cli/pkg/platform/model"
+	"github.com/ActiveState/cli/pkg/project"
+	"github.com/machinebox/graphql"
+	"github.com/skratchdot/open-golang/open"
+	"gopkg.in/yaml.v3"
+)
+
+type Params struct {
+	Name         string
+	Version      string
+	Namespace    string
+	Description  string
+	Authors      captain.UsersFlag
+	Depends      captain.PackagesFlag
+	Filepath     string
+	MetaFilepath string
+	Edit         bool
+	Editor       bool
+}
+
+type Runner struct {
+	auth    *auth.Auth
+	out     output.Outputer
+	prompt  prompt.Prompter
+	project *project.Project
+	client  *gqlclient.Client
+}
+
+type primeable interface {
+	primer.Outputer
+	primer.Auther
+	primer.Projecter
+	primer.Prompter
+}
+
+func New(prime primeable) *Runner {
+	client := gqlclient.NewWithOpts(
+		api.GetServiceURL(api.ServiceBuildPlan).String(), 0,
+		graphql.WithHTTPClient(http.DefaultClient),
+		graphql.UseMultipartForm(),
+	)
+	client.EnableDebugLog()
+	client.SetTokenProvider(prime.Auth())
+	client.EnableDebugLog()
+	return &Runner{auth: prime.Auth(), out: prime.Output(), prompt: prime.Prompt(), project: prime.Project(), client: client}
+}
+
+func (r *Runner) Run(params *Params) error {
+	if !r.auth.Authenticated() {
+		return locale.NewInputError("err_auth_required")
+	}
+
+	if !fileutils.FileExists(params.Filepath) {
+		return locale.NewInputError("err_uploadingredient_file_not_found", "File not found: {{.V0}}", params.Filepath)
+	}
+
+	reqVars := request.PublishVariables{}
+
+	// Pass input from meta file
+	if params.MetaFilepath != "" {
+		if !fileutils.TargetExists(params.MetaFilepath) {
+			return locale.NewInputError("err_uploadingredient_file_not_found", "File not found: {{.V0}}", params.MetaFilepath)
+		}
+
+		b, err := fileutils.ReadFile(params.MetaFilepath)
+		if err != nil {
+			return locale.WrapInputError(err, "err_uploadingredient_file_read", "Could not read file: {{.V0}}", params.MetaFilepath)
+		}
+
+		if err := yaml.Unmarshal(b, &reqVars); err != nil {
+			return locale.WrapInputError(err, "err_uploadingredient_file_read", "Failed to unmarshal meta file, error received: {{.V0}}", err.Error())
+		}
+	}
+
+	// Namespace
+	if params.Namespace != "" {
+		reqVars.Namespace = params.Namespace
+	} else if reqVars.Namespace == "" {
+		reqVars.Namespace = model.NewSharedNamespace(r.project.Owner()).String()
+	}
+
+	// Validate namespace
+	prefix := r.project.Owner() + "/"
+	if !strings.HasPrefix(reqVars.Namespace, prefix) {
+		return locale.NewInputError("err_uploadingredient_namespace_invalid_org",
+			"Namespace should be prefixed '[ACTIONABLE]{{.V0}}[/RESET]', received: '[ACTIONABLE]{{.V1}}[/RESET]'.", prefix, reqVars.Namespace)
+	}
```

Comment:
This seems to not take care of non-personal orgs. Is that intentional? Org restrictions feel like something the backend should handle.

Reply:
What do you mean by this? I don't see anything that implies either personal or non-personal.

Reply:
Ah, yeah, nevermind. I read this to mean that only the current user org could be uploaded to, but it's the org that contains the specific project that is being uploaded to. Sorry about that.